src/test/ui/closures/2229_closure_analysis/run_pass/lit-pattern-matching-with-methods.rs

```diff
@@ -0,0 +1,24 @@
+//check-pass
+#![feature(capture_disjoint_fields)]
+//~^ WARNING: the feature `capture_disjoint_fields` is incomplete
+#![warn(unused)]
+#![feature(rustc_attrs)]
+#![feature(btree_drain_filter)]
+
+use std::collections::BTreeMap;
+use std::panic::{catch_unwind, AssertUnwindSafe};
+
+fn main() {
+    let mut map = BTreeMap::new();
+    map.insert("a", ());
+    map.insert("b", ());
+    map.insert("c", ());
+
+    {
+        let mut it = map.drain_filter(|_, _| true);
+        catch_unwind(AssertUnwindSafe(|| while it.next().is_some() {})).unwrap_err();
```

Comment:
`while it.next().is_some()` actually triggers the pattern matching code. I was having issues earlier because I wasn't borrowing the discr place when the PatKind was `Lit`.  I am gonna keep the test as is.

Reply:
Hmm interesting, I'd have thought that `next` being a method would completely capture `it` and then `is_some()` is again just a method call on something that goes chained on top. I guess these get desugared?